utils/writeable/src/write_or_ref.rs

```diff
@@ -0,0 +1,203 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::Writeable;
+use alloc::borrow::Cow;
+use alloc::string::String;
+use core::fmt;
+
+/// Bytes that have been partially validated as UTF-8 up to an offset.
+struct PartiallyValidatedUtf8<'a> {
+    // Safety Invariants:
+    // 1. The offset is less than or equal to the length of the slice.
+    // 2. The slice is valid UTF-8 up to the offset.
+    slice: &'a [u8],
+    offset: usize,
+}
+
+impl<'a> PartiallyValidatedUtf8<'a> {
+    fn new(slice: &'a [u8]) -> Self {
+        Self { slice, offset: 0 }
```

Comment:
nit: safety comment:

```rust
/// Safety: Field invariants maintained here trivially:
///   1. The offset 0 is ≤ all possible lengths of slice
///   2. The slice contains nothing up to the offset zero
```


We typically want similar safety comments for whenever one of the fields is mutated.